common/thorhelper/thorsoapcall.cpp

```diff
@@ -389,27 +389,47 @@ class BlackLister : public CInterface, implements IThreadFactory
             throw MakeStringException(-1, "blacklisted socket %s", s.str());
         }
         Owned<IException> exc;
-        try
+
+        unsigned numAttemptsRemaining = retries+1;
+        if (!numAttemptsRemaining)
+            numAttemptsRemaining++;
+
+        unsigned connectTimeMS = timeoutMS;
+        if (timeoutMS == WAIT_FOREVER)
+            connectTimeMS = 60000 / numAttemptsRemaining;
+
+        while(numAttemptsRemaining-- > 0)
         {
-            checkRoxieAbortMonitor(roxieAbortMonitor);
-            Owned<ISocket> sock;
-            Owned<ISocketConnectWait> scw = nonBlockingConnect(ep, timeoutMS == WAIT_FOREVER ? 60000 : timeoutMS*(retries+1));
-            for (;;)
+            try
             {
-                sock.setown(scw->wait(1000));//throws if connect fails or timeoutMS
                 checkRoxieAbortMonitor(roxieAbortMonitor);
-                if (sock)
-                    return sock.getLink();
+                Owned<ISocket> sock;
+                Owned<ISocketConnectWait> scw = nonBlockingConnect(ep, connectTimeMS);
+                for (;;)
+                {
+                    sock.setown(scw->wait(1000)); //throws if connect fails or timeoutMS
+                    checkRoxieAbortMonitor(roxieAbortMonitor);
+                    if (sock)
+                        return sock.getLink();
+                }
+            }
+            catch (IJSOCK_Exception *e)
+            {
+                // MCK - do we checkTimeLimitExceeded(&remainingMS) and possibly error out if timelimit exceeded ?
+                if (numAttemptsRemaining > 0)
+                {
+                    e->Release();
+                }
+                else
+                {
+                    EXCLOG(e,"BlackLister::connect");
```

Comment:
I suspect this may cause confusion in the logs - the fact that the class that tried to do the connect is called "Blacklister" is an implementation detail, but we don't want users looking at logs to think that the blacklister is the cause of the failed connection (we've been down that road before). It's just the messenger...

Reply:
ok, good point, I will change this log msg to not have "Blacklister"